### PR TITLE
Add MaxBin2 as alternative binner

### PR DESCRIPTION
**Pull request by jfy133**
Adds MaxBin as an alternative binner as proposed in #252 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jfy133** [comment]
Hi @d4straub This is pretty much ready.

I only have two outstanding 'issues' which I'm somewhat unable to solve myself:

1. **Major**: I'm unable to get the test to run completely, because BUSCO_PLOT fails exit code 1 but with no error messages nor logging information. If I run `.command.sh` manually, it works fine.  I wonder if it is related to the complicated `bash` stuff... but I think I need a second pair of eye son that one.
2. **Minor**: `split_fasta.py` makes output file names by stripping off the last two extensions (i.e. anything after the last two `.`s), howevre the _unbinned_ files from MaXbin2 ends just in `.noclass.gz`, rather than `.noclass.fa.gz`. This means the final output files for the unbinned files are not so clearly defined as unbinned compraed to standard contigs (even if in an `unbinned` results folder). if you could tweak 'spit_fasta' to just remove `.{fa,fasta,fna}.gz` or similar? (I'm not a pyhton person sorry...)


help on either would be appreciated :pray:

File: subworkflows/local/binning.nf [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        .set{ ch_binning_results_final }
+
+    emit:
+    bins                                         = ch_binning_results_final
+    unbinned                                     = SPLIT_FASTA.out.unbinned
+    tooshort                                     = METABAT2_METABAT2.out.tooshort
+    lowdepth                                     = METABAT2_METABAT2.out.lowdepth
+    depths_summary                               = MAG_DEPTHS_SUMMARY.out.summary
+    metabat2_version                             = METABAT2_METABAT2.out.versions
+    metabat2_jgisummarizebamcontigdepths_version = METABAT2_JGISUMMARIZEBAMCONTIGDEPTHS.out.versions
```
Comment (d4straub) [inline subworkflows/local/binning.nf:152]: Is the Maxbin2 version reported somewhere? I would have thought here it should appear (and in `mag.nf` where I also cannot finnd it)

**Review by d4straub — COMMENTED** [review]
Looks good.
- about `BUSCO_PLOT`: could it be related to the warning `For 2 bin(s) BUSCO did not find any matching genes:`?
- about `split_fasta.py`: also not a python person, would it be solvable by not modifying `split_fasta.py` but instead using a suffix for maxbin2 output?

I have no time this week to look into that files in more detailed, sorry. Maybe next week.

File: docs/output.md [hunk 2]
```diff
@@ -253,8 +253,8 @@ Sequencing depth per contig and sample is generated by `jgi_summarize_bam_contig
 <summary>Output files</summary>
 
 * `GenomeBinning/MetaBAT2/`
-    * `bins/[assembler]-[sample/group].*.fa.gz`: Genome bins retrieved from input assembly
-    * `unbinned/[assembler]-[sample/group].unbinned.[1-9]*.fa.gz`: Contigs that were not binned with other contigs but considered interesting. By default, these are at least 1 Mbp (`--min_length_unbinned_contigs`) in length and at most the 100 longest contigs (`--max_unbinned_contigs`) are reported
+    * `bins/[assembler]-[binner]-[sample/group].*.fa.gz`: Genome bins retrieved from input assembly
+    * `unbinned_filtered/[assembler]-[binner]-[sample/group].unbinned.[1-9]*.fa.gz`: Contigs that were not binned with other contigs but considered interesting. By default, these are at least 1 Mbp (`--min_length_unbinned_contigs`) in length and at most the 100 longest contigs (`--max_unbinned_contigs`) are reported
```
[the remark this answers is not in the document]
Comment (jfy133) [reply docs/output.md:257]: Ah fair point. Sorry, this was a left over when I initially was trying to put the output from two modules into one, and apparently `publishDir` doesn't like this (will overwrite). I forgot to change this back. Will rename it back to `unbinned`

File: conf/modules.config [hunk 3]
```diff
@@ -350,34 +350,51 @@ process {
...
+            [
+                path: { "${params.outdir}/GenomeBinning/MaxBin2/discarded" },
+                mode: 'copy',
+                pattern: '*.tooshort.gz'
+            ],
+        ]
+        ext.prefix = { "${meta.assembler}-MaxBin2-${meta.id}" }
+        // if no gene found, will crash so allow ignore so rest of pipeline
+        // completes but without MaxBin2 results
+        errorStrategy = { task.exitStatus in [ 255 ] ? 'ignore' : 'retry' }
```
[the remark this answers is not in the document]
Comment (jfy133) [reply conf/modules.config:377]: Readded

File: subworkflows/local/binning.nf (hunk 1 above)
Comment (jfy133) [reply subworkflows/local/binning.nf:152]: Oops, good Catch"

**Comment by jfy133** [comment]
> about BUSCO_PLOT: could it be related to the warning For 2 bin(s) BUSCO did not find any matching genes:?
about split_fasta.py: also not a python person, would it be solvable by not modifying split_fasta.py but instead using a suffix for maxbin2 output?

1. that's the strange thing, when I run `.command.sh` manually I do get (working!) output of BUSCO_PLOT (even if it is 'empty'

image

so this doesn't explain why the BUSCO_PLOT module is failing during the test run.

2. Unfortunately this doesn't work as MaxBin2 adds [0-90-90-9] at teh end of the prefix to indicate each bin. Maybe I can ask a pythonista colleague for a suggestion

**Comment by d4straub** [comment]
Unfortunately I have the same issue with `NFCORE_MAG:MAG:BUSCO_QC:BUSCO_PLOT` when I run the test locally. It doesnt fail for all test samples but only some, e.g. `SPAdes-test_minigut_sample2` and `SPAdes-test_minigut`. I am clueless though why that happens.

**Comment by jfy133** [comment]
@d4straub I think I have it 'working' now. Slight fail because two of the unbinned contigs not having enoug hgene content form MaxBin2. But otherwise should be OK. Could you check you see the correct output, as you expect?

image

**Comment by d4straub** [comment]
I run the test profile, looks good, except:
- I expected the content of `results/GenomeBinning/MetaBAT2/unbinned/discarded` in `results/GenomeBinning/MetaBAT2/discarded`, any reason to have twice "discarded"? I see that you also changed that in the output docu, so that seems intended.
- MAGs in `results/GenomeBinning/MetaBAT2/unbinned` are not processed by Prokka (maybe also not the MaxBin2 unbinned?), but they should (`results/GenomeBinning/bin_summary.tsv` has all expected unbinned & unbinned entries, as far as I can see)

**Comment by jfy133** [comment]
> * I expected the content of `results/GenomeBinning/MetaBAT2/unbinned/discarded` in `results/GenomeBinning/MetaBAT2/discarded`, any reason to have twice "discarded"? I see that you also changed that in the output docu, so that seems intended.

That was to separate out those those 'discarded' by split_fasta  (in the former), and and those 'discarded' by the binner itself (in the latter case). I was personally finding it difficult to follow which was which at different points. 

> * MAGs in `results/GenomeBinning/MetaBAT2/unbinned` are not processed by Prokka (maybe also not the MaxBin2 unbinned?), but they should (`results/GenomeBinning/bin_summary.tsv` has all expected unbinned & unbinned entries, as far as I can see)

Oops missed that! Will push that now

**Comment by d4straub** [comment]
Sorry, found another problem, now neither Metabat2 nor Maxbin2 versions are appearing in the multiQC report.

**Comment by jfy133** [comment]
I'm getting some large python traceback in the DUMPSOFTWAREVERSIONS module , so this will take a bit longer (not really sure what's causing this stil), sorry!

**Review by d4straub — APPROVED** [review]
Looks good!

**Comment by jfy133** [comment]
Thanks! DAS_Tool should come early next week

**Pull request merged**